### DISCUSSION TITLE
Guidance for new contributors

### DISCUSSION THREAD

**Discussion by m-holger** [m-holger/pdfarranger#10] [Ideas]
Place holder

**Comment by m-holger** [comment]
@dreua

> Loosing the About Box discussion is unfortunate, I think it could have become valuable information for other new contributors. Is there any way to restore it or do you have a copy that you could mail me? I could give it a quick edit and put it in a wiki page for preservation.

I don't think the discussion itself would be useful as it is too disorganised, and therefore I do not see a point trying to restore it. I agree that there should be an idiot's guide for new contributors, and that some of the info from the discussion should go into it. However, I don't think that this would be enough to address the problems that new contributors are facing, some of which are:


#### 1 Forum for asking questions

There is no obvious place to ask questions. You volunteered your assistance in response to an me opening https://github.com/pdfarranger/pdfarranger/issues/439#issue-837079837 (where I volunteered to have a stab at the implementation), but
- I don't consider that what I suggested was an issue (more of an observation), and therefore might not have opened it.
- Seeing that there is no obvious way of asking questions about PDF Arranger, I might very well not have volunteered.   

Besides, asking questions effectively off-line (be it in a private discussion or buried in an issue, which I would consider off-topic) has the following disadvantages:
- It is probably not the most effective use of a team members time, as most of the basic questions probably could be answered just as well by the community
- Questions and answers are not visible to others, and therefore the same questions will be asked repeatedly
- There is no way to identify common questions and collect them in a FAQ - often the most important and helpful questions are the ones you don't realise you have 

Would it make sense to enable "discussions" on GitHub or provide some alternative forum?

(As an aside, I think it would be easier for users investigating possible bugs if the issues list was limited to genuine issues; and feature requests and discussions of possible new feature were moved away to a separate forum)

#### 2 Coding standards

There are a number of different coding styles used in the application, probably for historic reasons. While I can guess for some of them what style is desired, it is a bit daunting modifying code based on a guess.


#### 3 Documentation

I think it is fair to say that, for reasons that I fully understand, the developer documentation is not as good as one might hope for. To give an outsider a reasonable chance to contribute without excessive effort, I believe at a minimum there should be a high-level overview of the architecture.


#### 4 Vision, priorities and road map

It is pretty difficult for outsiders to contribute meaningfully if they don't know where the team members want to go with PDF Arranger and what they are working on.


I would be quite happy to draft some of the items (for starters,the idiots guide), but to make it worthwhile I would need access to and the support of the team members.

(Feel free to include others in the discussion as you see fit)

**Reply by dreua** [reply]
> I don't think the discussion itself would be useful as it is too disorganised

Yes it was disorganized, that's why I wanted to give it a quick edit and put it in a Wiki page. It would be nice to have at least access to my own comments so that I can copy and paste stuff from there when similar questions come up.

### 2 Coding standards
I wasn't aware we had a number of different styles, those should be ideally standardized and the preferred style documented, maybe even CI-checked. I think PEP8 is a good base line but it is not necessary to follow it to the smallest detail. If in doubt look at PEP8. And don't be too scared or daunted, we will happily explain that stuff in a PR. That's the point of review, the goal is not to open a PR with perfect commits, if it were perfect you could push to stable right away.

### Goals
I think we'll need to discuss this internally. The big goal was and is to have a maintained version of PDF-Shuffler. I think we achieved that quite good. Other than that I think there is some difference between the team members and it can also vary over time for each individual. I happened to have some spare time and felt like helping you, but this might change soon if I start to work a full time job.

**Reply by m-holger** [reply]
@dreua I've emailed you what I've got.

I'll leave you to discuss internally.

**Reply by dreua** [reply]
Your email didn't arrive

**Reply by m-holger** [reply]
Not sure what happened. Probably just my understanding of GitHub


Your emails:



Hey,

    No we don't have any target release dates that I know of. No pressure, it would be cool to have this for 1.8.0 but if you want to do it and it takes longer then it goes into the next release after that. Maybe we need better communication of what scheduling stuff for releases means, for me in this case it just meant that I think it's a really good idea :)
    Quite sure that wasn't a deliberate choice, feel free to make that a link!

Yeah, taking these questions offline is absolutely fine with me. Just remember to tag me or else I won't see them. You could link them in the main issue just in case someone else is interested to follow along, but I leave that decision up to you.

Thank you for taking care of this, happy to follow you along and help where I can. Feel free to tag me in your commits, too!


Yeah, well translations can be an annoying topic if you ask me, I'm not so sure about each and every detail myself but here is what I would do:

    Please don't update existing strings just to use "f"-strings. I'm quite sure it will break translations unnecessarily. If you add new strings or the string changes anyway we could think about f-strings, but I don't want to decide this on my own. I'd say if there is no strong need for your feature to use f-strings just stick with the %s. Introduction or switching to f-strings should be discussed in a new issue with the whole team.
    Yes you are right that it won't be translated. I haven't thought of the danger of a runtime error before but I think you are right too: As the string changes and becomes untranslated it should not be a problem. I can easily test that on linux (something like LANG=de_DE pdfarranger), could you remind me to do that once you open a PR?
    When a significant amount of string changes happened, Jerome tagged all previous translators to update translations. I guess we will do that again if it becomes necessary. For your feature I wouldn't worry too much about that, its only the About dialog, so no critical feature like "delete" or "overwrite". Furthermore, people will need to speak or translate to English anyway in order to communicate with us. It is acceptable to have untranslated strings here imo.
    No, you don't need to provide any translations at all and you should not attempt to translate to languages that you don't speak fluently.

Good questions, let me no if something unclear :)


Thanks for the PR!

Please link to the corresponding GH issue so everyone involved can easily navigate between the two. (Ideally add the full link to the commit message for future reference. Use Closes: to automatically close the issue once the commit is merged.)

Two minor nitpicks: The two commits should be squashed into one. You may use the force (push -f) on your feature branch, it should automatically update this PR. And commit messages should be present tense (Add instead of Added).


I think kbengs is referring to the general convention. Not sure we follow and enforce this in all cases within PDF Arranger but it is good practice to follow it

https://www.google.com/search?q=git%20commit%20message%20convention&ie=utf-8&oe=utf-8&client=firefox-b-m



This Google Link is not that great, here is the gist:

Rules for a great git commit message style

    Separate subject from body with a blank line
    Do not end the subject line with a period
    Capitalize the subject line and each paragraph
    Use the imperative mood in the subject line
    Wrap lines at 72 characters
    Use the body to explain what and why you have done something. In most cases, you can leave out details about how a change has been made.

Source https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53


This is all fine, thank you! I prefer to use the full Github URL (https://github.com/pdfarranger/pdfarranger/issues/439 instead of just #439) in the commit message because it is clickable outside Github but that is a really matter of personal taste.

The Codecov report looked good, but you can simply ignore it.

**Reply by dreua** [reply]
Thx
